src/trace_ng/log.zig

```diff
@@ -0,0 +1,1048 @@
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logfWithFields(level, fmt, args, fields),
+            }
+        }
+    };
+}
+
+pub const Logger = ScoppedLogger(null);
+
+/// An instance of `ScopedLogger` that logs to the standard err.
+pub fn StandardErrLogger(comptime scope: ?[]const u8) type {
+    return struct {
+        const Self = @This();
+        max_level: Level,
+        exit_sig: std.atomic.Value(bool),
+        allocator: Allocator,
+        log_allocator: Allocator,
+        log_allocator_state: *RecycleFBA(.{}),
+        max_buffer: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) !*Self {
+            const recycle_fba = try config.allocator.create(RecycleFBA(.{}));
+            const max_buffer = config.max_buffer orelse return error.MaxBufferNotSet;
+            recycle_fba.* = try RecycleFBA(.{}).init(config.allocator, max_buffer);
+            const self = try config.allocator.create(Self);
+            self.* = .{
+                .allocator = config.allocator,
+                .log_allocator = recycle_fba.allocator(),
+                .log_allocator_state = recycle_fba,
+                .max_buffer = max_buffer,
+                .exit_sig = AtomicBool.init(false),
+                .max_level = config.max_level,
+                .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+                .handle = try std.Thread.spawn(.{}, run, .{self}),
+            };
+            return self;
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.channel.close();
+            if (self.handle) |*handle| {
+                std.time.sleep(std.time.ns_per_ms * 5);
+                self.exit_sig.store(true, .seq_cst);
+                handle.join();
+            }
+            self.channel.deinit();
+            self.log_allocator_state.deinit();
+            self.allocator.destroy(self.log_allocator_state);
+            self.allocator.destroy(self);
+        }
+
+        pub fn unscoped(self: Self) Logger {
+            return .{
+                .allocator = self.allocator,
+                .recycle_fba = self.log_allocator_state,
+                .max_buffer = self.max_buffer,
+                .max_level = self.max_level,
+                .exit_sig = self.exit_sig,
+                .channel = self.channel,
+                .handle = self.handle,
+            };
+        }
+
+        pub fn withScope(self: Self, comptime new_scope: anytype) ScoppedLogger(new_scope) {
+            return .{
+                .allocator = self.allocator,
+                .recycle_fba = self.log_allocator_state,
+                .max_buffer = self.max_buffer,
+                .max_level = self.max_level,
+                .exit_sig = self.exit_sig,
+                .channel = self.channel,
+                .handle = self.handle,
+            };
+        }
```

Comment:
These functions would fail to compile if they were actually used. `ScoppedLogger` is a union that holds only a pointer to the logger implementation.

For these to work it just needs to create the union from the pointer to self. Also I might change the function names to clarify that it's returning a different type. This is consistent with the interface pattern used in zig std.

```suggestion
        pub fn logger(self: *Self) Logger {
            return .{ .standard = self };
        }

        pub fn scopedLogger(self: Self, comptime new_scope: anytype) ScoppedLogger(new_scope) {
            return .{ .standard = self };
        }
```